js/reference.js

```diff
@@ -1164,26 +1171,89 @@ var ERMrest = (function(module) {
                 };
-
+                var ownReference = this;
                 this._server._http.delete(this.uri, config).then(function deleteReference(response) {
-
                     defer.resolve();
                 }, function error(response) {
-                    var error = module._responseToError(response);
-                    return defer.reject(error);
+                    var status = response.status || response.statusCode;
+                    // If 412 Precondition Failed it means that ETags don't match
+                    if (status == 412) {
+                        var MAX_TRIES = 3, numTriesLeft = MAX_TRIES;
+                        // Check if the record still exists. If it does, compare the data.
+                        // If it doesn't, the data has already been deleted. If the
+                        // .read goes to error callback, then retry the read.
+                        var readReference = function readReference(ref) {
+                            ref.read(tuples.length).then(function getPage(page) {
+                                var oldData = tuples, currentData = page.tuples, mismatchFound = false;
+                                // If the referenced rows have already been deleted, page.tuples is an empty array.
+                                // Resolve the promise successfully like normal.
+                                if (currentData.length === 0) {
+                                    return defer.resolve();
+                                }
+
+                                // Else, compare the data.
+                                if (currentData.length !== oldData.length) {
+                                    mismatchFound = true;
+                                }
+
+                                if (!mismatchFound) {
+                                    // If findMismatch is true, this
+                                    // loop breaks and the loop returns true.
+                                    var findMismatch = function findMismatch(key) {
+                                        return oldTuple[key] !== currentTuple[key];
+                                    };
+                                    for (var i = 0, len = oldData.length; i < len; i++) {
+                                        var oldTuple = oldData[i]._data, currentTuple = currentData[i]._data;
+                                        mismatchFound = Object.keys(oldTuple).some(findMismatch);
+                                        if (mismatchFound) {
+                                            break;
+                                        }
+                                    }
+                                }
+                                // If old data matches current data, retry the delete w/ updated ETag
+                                if (!mismatchFound && numTriesLeft > 0) {
+                                    numTriesLeft--;
+                                    return ref.delete(tuples);
```

Comment:
Should this be `currentData`? If not, it doesn't look like you update the `ETag` on `tuples` here before trying to delete again.